### PR TITLE
`spack ci`: Don't require OIDC initialization for noop

### PR DESCRIPTION
**Pull request by kwryankrattiger**
ref. https://github.com/spack/spack-infrastructure/pull/957

### CODE REVIEW

File: lib/spack/spack/schema/ci.py [hunk 1]
```diff
@@ -47,7 +47,7 @@
         "tags": {"type": "array", "items": {"type": "string"}},
         "variables": {
             "type": "object",
-            "patternProperties": {r"[\w\d\-_\.]+": {"type": "string"}},
+            "patternProperties": {r"[\w\d\-_\.]+": {"type": ["string", "number"]}},
```
Comment (kwryankrattiger) [inline lib/spack/spack/schema/ci.py:50]: @scottwittenburg I have been annoyed by this for ages and this finally required changing it, any reason this shouldn't work that you remember?

**Comment by kwryankrattiger** [comment]
noop is passing with this PR: https://gitlab.spack.io/spack/spack/-/jobs/12945613

**Comment by kwryankrattiger** [comment]
Note, since this exits 0 from the OIDC setup the rest of noop script is completely skipped. I think this is fine for our purposes, but worth noting.

File: lib/spack/spack/schema/ci.py (hunk 1 above)
Comment (scottwittenburg) [reply lib/spack/spack/schema/ci.py:50]: I can't think of any reason why this would be a problem.

**Review by scottwittenburg — COMMENTED** [review]
> Note, since this exits 0 from the OIDC setup the rest of noop script is completely skipped. I think this is fine for our purposes, but worth noting.

I don't see any documentation stating why this would be the case.  But I see in the linked job that what you noted seems to be true.  I wonder if gitlab treats the `pre_build_script` like the `before_script`, and gloms it all together to make the step script that gets run? 

At any rate, I agree this shouldn't affect any other projects other than spack ci on gitlab.spack.io, since that variable would just be ignored.  And it seems fine for our purposes to short-circuit the no-op job.  However, maybe it's just the naming of the variable (given the observed behavior), but it seems weird that just not requiring OIDC means a job is effectively useless.   Maybe make a comment in the no-op job definition stating that `CI_OIDC_REQUIRED` isn't generally useful because of the behavior you observed.

~One last thing, and this seems key: are you sure that, as it stands, this won't short-circuit all of our jobs? My bash skills aren't great, but is it possible that we might need to add `CI_OIDC_REQUIRED: 1` to all the other jobs to avoid that situation?~ Edit: Maybe nevermind, I guess if the infra PR is merged, that would already be happening.

**Comment by kwryankrattiger** [comment]
> I wonder if gitlab treats the pre_build_script like the before_script, and gloms it all together to make the step script that gets run?

I think that is how it is done, which makes sense since environment variables are shared between `pre_build_script`, `before_script`, and `script`.

>  However, maybe it's just the naming of the variable (given the observed behavior), but it seems weird that just not requiring OIDC means a job is effectively useless.

I agree it is kind of odd, We could make this better in the CI infra change. I am not sure if this is the right place to document that or in the infrastructure somewhere since it is specific to our runners with the `service` tag. However, in our cases, if you aren't using OIDC then the job can't actually doing anything at all since all of our CI operations need S3 access.

**Comment by scottwittenburg** [comment]
> I am not sure if this is the right place to document that or in the infrastructure somewhere since it is specific to our runners with the `service` tag.

In my mind, the point of documenting it here is to prevent someone from trying to define a new job down the road that doesn't need oidc.  But I'll leave that up to you.

**Review by scottwittenburg — APPROVED** [review]

**Comment by kwryankrattiger** [comment]
noop runs script now, skips all of the OIDC setup.

 https://gitlab.spack.io/spack/spack/-/jobs/12950358

**Pull request merged**